platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dsl/dependencies/GradlePluginVariantsResolutionFailureDescriber.java

```diff
@@ -0,0 +1,101 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.artifacts.dsl.dependencies;
+
+import org.gradle.api.attributes.AttributeContainer;
+import org.gradle.api.attributes.plugin.GradlePluginApiVersion;
+import org.gradle.api.internal.DocumentationRegistry;
+import org.gradle.internal.component.ResolutionFailure;
+import org.gradle.internal.component.ResolutionFailureDescriber;
+import org.gradle.internal.component.NoMatchingGraphVariantsException;
+import org.gradle.internal.component.ResolutionCandidateAssessor.AssessedCandidate;
+import org.gradle.util.GradleVersion;
+
+import java.util.List;
+import java.util.Optional;
+
+public class GradlePluginVariantsResolutionFailureDescriber implements ResolutionFailureDescriber<NoMatchingGraphVariantsException> {
+    private static final String GRADLE_VERSION_TOO_OLD_TEMPLATE = "Plugin %s requires at least Gradle %s (this build used %s).";
+    private static final String NEEDS_NEWER_GRADLE_SECTION = "sub:updating-gradle";
+
+    private final DocumentationRegistry documentationRegistry;
+    private final GradleVersion currentGradleVersion;
+
+    protected GradlePluginVariantsResolutionFailureDescriber(DocumentationRegistry documentationRegistry, GradleVersion currentGradleVersion) {
+        this.documentationRegistry = documentationRegistry;
+        this.currentGradleVersion = currentGradleVersion;
+    }
+
+    @Override
+    public boolean canDescribeFailure(ResolutionFailure failure) {
+        return failure.getType() == ResolutionFailure.ResolutionFailureType.NO_MATCHING_VARIANTS && isPluginRequestUsingApiVersionAttribute(failure.getRequestedAttributes());
+    }
+
+    @Override
+    public NoMatchingGraphVariantsException describeFailure(ResolutionFailure failure) {
+        Optional<String> minRequiredGradleVersion = findHigherRequiredVersionOfGradle(failure.getCandidates());
+        if (minRequiredGradleVersion.isPresent()) {
+            String message = buildPluginNeedsNewerGradleVersionFailureMsg(failure.getRequestedName(), minRequiredGradleVersion.get());
+            NoMatchingGraphVariantsException result = new NoMatchingGraphVariantsException(message);
+            suggestUpdateGradle(result, minRequiredGradleVersion.get());
+            suggestDowngradePlugin(result, failure.getRequestedName());
+            return result;
+        } else {
+            throw new IllegalStateException("Expected to find a higher required version of Gradle.");
```

Comment:
When would we ever hit this?  I guess if none of the candidates had a GRADLE_PLUGIN_API_VERSION_ATTRIBUTE?

Reply:
Which is why it's illegal 😉.  I need to look at cleaning this code up, this is mostly cut and paste from my first attempt doing this in the `ResolutionFailureHandler` itself.